File: src/gcs/newton.py
"""Our own least-squares solvers: Powell's DogLeg (default) and Levenberg–Marquardt.

Both minimise ½‖r(z)‖² given callbacks for r(z) and J(z).  The Gauss–Newton
step is the *minimum-norm* least-squares solution of J p = −r, so
under-constrained sketches (the normal case while editing) move as little as
possible — least-change behaviour is what users expect from dragging.

Dense path (n_free ≤ DENSE_MAX): J as ndarray, LAPACK lstsq (SVD) — also gives
the numerical rank.  Sparse path: CSR J, LSMR from a zero start (min-norm in
the limit).  Reference: Nocedal & Wright ch. 4 & 10; PlaneGCS's DogLeg.
"""

from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

import numpy as np
import scipy.sparse as sp
import scipy.sparse.linalg as spla
from scipy.linalg import lapack

Vec = Any  # npt.NDArray[np.float64]; kept loose so callbacks may return csr or ndarray


@dataclass
class Info:
    status: int          # 0 = residual below tol, 1 = step below xtol, 2 = gradient below gtol,
    #                      3 = trust region collapsed, 4 = max iterations, -1 = failed
    message: str
    nfev: int
    njev: int
    iterations: int
    rank: int | None     # numerical rank of J at the solution (dense path only)


_MSG = {0: "residual tolerance reached", 1: "step size below xtol", 2: "gradient below gtol",
        3: "trust region collapsed", 4: "max iterations reached", -1: "failed"}


_RCOND = 1e-12
_lwork_cache: dict[tuple[int, int], int] = {}


def min_norm_lstsq(J: Vec, b: Vec, rcond: float = _RCOND) -> tuple[Vec, int]:
    """Minimum-norm least-squares solution of J p = b via LAPACK dgelsy (rank-revealing
    QR with column pivoting).  Returns (p, numerical rank).  ~6× faster than
    np.linalg.lstsq (SVD) at sketch sizes and gives the rank Stage 2/4 need."""
    m, n = J.shape
    key = (m, n)
    lwork = _lwork_cache.get(key)
    if lwork is None:
        lwork = _lwork_cache[key] = int(lapack.dgelsy_lwork(m, n, 1, rcond)[0].real)
    bb = np.zeros((max(m, n), 1))     # LAPACK wants ldb >= max(m, n)
    bb[:m, 0] = b
    _, x, _, rank, info = lapack.dgelsy(J, bb, np.zeros(n, dtype=np.int32), rcond, lwork)
    if info != 0:
        raise np.linalg.LinAlgError(f"dgelsy failed: info={info}")
    return x[:n, 0], int(rank)


def _norm(v: Vec) -> float:
    return math.sqrt(float(v @ v))


def _absmax(v: Vec) -> float:
    return float(np.max(np.abs(v))) if v.size else 0.0


_EPS_REL = 1e-12


def sparse_gn_step(J: Any, r: Vec) -> Vec:
    """Sparse Gauss–Newton step from the regularized normal equations
    (JᵀJ + εI) p = −Jᵀr, factored with SuperLU.  ε = 1e-12·max diag keeps
    rank-deficient (under-constrained) systems solvable and the step within
    ~1e-6 relative of the true minimum-norm one — LSMR/LSQR are 30–50× slower
    here and dense QR is O(n³).  (Sparse QR — SPQR — is the Stage-1 C option.)"""
    A = (J.T @ J).tocsc()
    eps = _EPS_REL * float(A.diagonal().max(initial=0.0)) or 1e-30
    A = A + sp.identity(A.shape[0], format="csc") * eps
    return spla.splu(A).solve(-(J.T @ r))


def _gn_step(J: Any, r: Vec) -> tuple[Vec, int | None]:
    """Gauss–Newton step p solving J p ≈ −r (minimum-norm on the dense path)."""
    if not isinstance(J, np.ndarray):   # sparse (sp.issparse is an ABC check — 25 µs)
        return sparse_gn_step(J, r), None
    return min_norm_lstsq(J, -r)


def dogleg(
    fun: Callable[[Vec], Vec],
    jac: Callable[[Vec], Any],
    z0: Vec,
    *,
    ftol: float = 1e-12,      # stop when max|r| < ftol (absolute; caller scales by sketch extent²)
    xtol: float = 1e-12,      # stop when ‖p‖ < xtol·(1 + ‖z‖)
    gtol: float = 1e-14,      # stop when ‖Jᵀr‖∞ < gtol
    max_iter: int = 100,
    max_nfev: int | None = None,
) -> tuple[Vec, Info]:
    z = np.array(z0, dtype=np.float64)
    r = fun(z)
    nfev, njev = 1, 0
    delta = np.inf            # first step is the full Gauss–Newton step
    rank: int | None = None
    max_nfev = max_nfev if max_nfev is not None else 4 * max_iter
    for it in range(max_iter):
        if _absmax(r) < ftol:
            return z, Info(0, _MSG[0], nfev, njev, it, rank)
        J = jac(z)
        njev += 1
        f = 0.5 * float(r @ r)
        g = J.T @ r
        gnorm = _absmax(g)
        if gnorm < gtol:
            return z, Info(2, _MSG[2], nfev, njev, it, rank)
        p_gn, rank = _gn_step(J, r)
        gn_norm = _norm(p_gn)
        # -- choose the dogleg step inside the trust region --
        if gn_norm <= delta:
            p = p_gn
        else:
            Jg = J @ g
            alpha = float(g @ g) / float(Jg @ Jg)
            p_sd = -alpha * g
            sd_norm = _norm(p_sd)
            if sd_norm >= delta:
                p = p_sd * (delta / sd_norm)
            else:  # on the segment p_sd → p_gn where ‖p‖ = delta
                d = p_gn - p_sd
                a, b, c = float(d @ d), 2 * float(p_sd @ d), sd_norm**2 - delta**2
                tau = (-b + np.sqrt(b * b - 4 * a * c)) / (2 * a)
                p = p_sd + tau * d
        pnorm = _norm(p)
        if pnorm < xtol * (1 + _norm(z)):
            return z, Info(1, _MSG[1], nfev, njev, it, rank)
        # -- evaluate and update the trust region --
        z_new = z + p
        r_new = fun(z_new)
        nfev += 1
        f_new = 0.5 * float(r_new @ r_new)
        lin = r + J @ p
        pred = f - 0.5 * float(lin @ lin)
        rho = (f - f_new) / pred if pred > 0 else (1.0 if f_new < f else -1.0)
        if rho > 0:
            z, r = z_new, r_new
            if rho > 0.75:
                delta = max(delta, 3 * pnorm) if np.isfinite(delta) else np.inf
            elif rho < 0.25:
                delta = 0.5 * pnorm
        else:
            delta = 0.25 * pnorm
        if delta < 1e-15 * (1 + _norm(z)):
            return z, Info(3, _MSG[3], nfev, njev, it + 1, rank)
        if nfev >= max_nfev:
            break
    return z, Info(4, _MSG[4], nfev, njev, max_iter, rank)


def levenberg_marquardt(
    fun: Callable[[Vec], Vec],
    jac: Callable[[Vec], Any],
    z0: Vec,
    *,
    ftol: float = 1e-12,
    xtol: float = 1e-12,
    gtol: float = 1e-14,
    max_iter: int = 100,
    max_nfev: int | None = None,
    tau: float = 1e-8,
) -> tuple[Vec, Info]:
    """LM with Nielsen's damping update; solves (JᵀJ + λD)p = −Jᵀr with D = diag(JᵀJ)
    (floored so unconstrained parameters stay well-posed).  Small initial τ: sketch
    solves are zero-residual problems started from a warm start, so the first steps
    should be nearly Gauss–Newton."""
    z = np.array(z0, dtype=np.float64)
    r = fun(z)
    nfev, njev = 1, 0
    max_nfev = max_nfev if max_nfev is not None else 4 * max_iter
    lam, nu = -1.0, 2.0
    for it in range(max_iter):
        if _absmax(r) < ftol:
            return z, Info(0, _MSG[0], nfev, njev, it, None)
        J = jac(z)
        njev += 1
        sparse = not isinstance(J, np.ndarray)
        A = (J.T @ J).tocsc() if sparse else J.T @ J
        g = J.T @ r
        if _absmax(g) < gtol:
            return z, Info(2, _MSG[2], nfev, njev, it, None)
        diag = np.asarray(A.diagonal()).ravel()
        dmax = float(diag.max(initial=0.0))
        D = np.maximum(diag, 1e-8 * dmax if dmax > 0 else 1e-8)
        if lam < 0:
            lam = tau * (dmax if dmax > 0 else 1.0)
        f = 0.5 * float(r @ r)
        while True:
            if sparse:
                p = spla.spsolve(A + sp.diags(lam * D), -g)
            else:
                p = np.linalg.solve(A + np.diag(lam * D), -g)
            pnorm = _norm(p)
            if pnorm < xtol * (1 + _norm(z)):
                return z, Info(1, _MSG[1], nfev, njev, it, None)
            z_new = z + p
            r_new = fun(z_new)
            nfev += 1
            f_new = 0.5 * float(r_new @ r_new)
            pred = 0.5 * float(p @ (lam * D * p - g))
            rho = (f - f_new) / pred if pred > 0 else -1.0
            if rho > 0:
                z, r = z_new, r_new
                lam *= max(1 / 3, 1 - (2 * rho - 1) ** 3)
                nu = 2.0
                break
            lam *= nu
            nu *= 2
            if nfev >= max_nfev or lam > 1e32:
                return z, Info(4 if nfev >= max_nfev else 3, _MSG[4 if nfev >= max_nfev else 3], nfev, njev, it + 1, None)
        if nfev >= max_nfev:
            break
    return z, Info(4, _MSG[4], nfev, njev, max_iter, None)

File: csrc/newton.c
/* Powell's DogLeg (default) and Levenberg-Marquardt, both minimising 0.5*||r(z)||^2.
 *
 * The Gauss-Newton step is the *minimum-norm* least-squares solution of J p = -r, so
 * under-constrained sketches (the normal case while editing) move as little as possible.
 * Dense path: the complete orthogonal decomposition in linalg.c, which also reports the
 * rank.  Sparse path: the regularized normal equations (J^T J + eps I) p = -g factored
 * by LDL^T, which keeps rank-deficient systems solvable.
 */
#include "gcs.h"
#include "sparse.h"
#include "system.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

#define EPS_REL 1e-12

static double vnorm(const double *v, int n)
{
    double s = 0.0;
    for (int i = 0; i < n; i++) s += v[i] * v[i];
    return sqrt(s);
}

static double vdot(const double *a, const double *b, int n)
{
    double s = 0.0;
    for (int i = 0; i < n; i++) s += a[i] * b[i];
    return s;
}

static double absmax(const double *v, int n)
{
    double mx = 0.0;
    for (int i = 0; i < n; i++) { double a = fabs(v[i]); if (a > mx) mx = a; }
    return mx;
}

/* -- the Jacobian, dense or sparse, behind one interface -------------------- */

typedef struct {
    gcs_system *s;
    int dense, m, n;
    double *J;        /* dense m*n */
    double *data;     /* csr values (points at s->csr_data) */
    double *work;     /* n scratch, sparse path only */
    int rank;
} jac_ctx;

static void jac_eval(jac_ctx *c, const double *z)
{
    if (c->dense) gcs_system_jacobian_dense(c->s, z, c->J);
    else gcs_system_csr_data(c->s, z, c->data);
}

static void res_eval(jac_ctx *c, const double *z, double *out)
{
    gcs_system_residuals(c->s, z, out);
}

/* out (n) <- J^T v (m) */
static void jt_mul(const jac_ctx *c, const double *v, double *out)
{
    memset(out, 0, sizeof(double) * (size_t)c->n);
    if (c->dense) {
        for (int i = 0; i < c->m; i++) {
            double vi = v[i];
            if (vi == 0.0) continue;
            const double *row = c->J + (size_t)i * c->n;
            for (int j = 0; j < c->n; j++) out[j] += row[j] * vi;
        }
    } else {
        const int32_t *ip = c->s->csr_indptr, *ii = c->s->csr_indices;
        for (int i = 0; i < c->m; i++) {
            double vi = v[i];
            if (vi == 0.0) continue;
            for (int p = ip[i]; p < ip[i + 1]; p++) out[ii[p]] += c->data[p] * vi;
        }
    }
}

/* out (m) <- J v (n) */
static void j_mul(const jac_ctx *c, const double *v, double *out)
{
    if (c->dense) {
        for (int i = 0; i < c->m; i++) {
            const double *row = c->J + (size_t)i * c->n;
            double s = 0.0;
            for (int j = 0; j < c->n; j++) s += row[j] * v[j];
            out[i] = s;
        }
    } else {
        const int32_t *ip = c->s->csr_indptr, *ii = c->s->csr_indices;
        for (int i = 0; i < c->m; i++) {
            double s = 0.0;
            for (int p = ip[i]; p < ip[i + 1]; p++) s += c->data[p] * v[ii[p]];
            out[i] = s;
        }
    }
}

/* p <- the Gauss-Newton step solving J p ~= -r (minimum norm on the dense path). */
static int gn_step(jac_ctx *c, const double *r, const double *g, double *p)
{
    if (c->dense) {
        double *A = (double *)malloc(sizeof(double) * (size_t)c->m * c->n);
        double *B = (double *)malloc(sizeof(double) * (size_t)(c->m ? c->m : 1));
        memcpy(A, c->J, sizeof(double) * (size_t)c->m * c->n);
        for (int i = 0; i < c->m; i++) B[i] = -r[i];
        c->rank = gcs_min_norm_lstsq(c->m, c->n, 1, A, B, 1e-12, p);
        free(A); free(B);
        return 0;
    }
    if (!c->s->ata) c->s->ata = gcs_ata_new(c->m, c->n, c->s->csr_indptr, c->s->csr_indices);
    gcs_ata *a = c->s->ata;
    gcs_ata_fill(a, c->data);
    gcs_ata_diag(a, c->work);
    double dmax = 0.0;
    for (int i = 0; i < c->n; i++) if (c->work[i] > dmax) dmax = c->work[i];
    double eps = EPS_REL * dmax;
    if (eps <= 0.0) eps = 1e-30;
    for (int i = 0; i < c->n; i++) { c->work[i] = eps; p[i] = -g[i]; }
    c->rank = -1;
    return gcs_ata_solve(a, c->work, p);
}

/* -- DogLeg ----------------------------------------------------------------- */

static int dogleg_core(jac_ctx *c, double *z, double *r, double ftol, double xtol, double gtol, int max_iter, int max_nfev,
                       gcs_info *info)
{
    int m = c->m, n = c->n;
    double *g = (double *)malloc(sizeof(double) * (size_t)n);
    double *p = (double *)malloc(sizeof(double) * (size_t)n);
    double *p_gn = (double *)malloc(sizeof(double) * (size_t)n);
    double *p_sd = (double *)malloc(sizeof(double) * (size_t)n);
    double *z_new = (double *)malloc(sizeof(double) * (size_t)n);
    double *r_new = (double *)malloc(sizeof(double) * (size_t)(m ? m : 1));
    double *tmp = (double *)malloc(sizeof(double) * (size_t)(m ? m : 1));
    int nfev = 1, njev = 0, status = 4, it = 0;
    double delta = INFINITY;
    c->rank = -1;
    for (it = 0; it < max_iter; it++) {
        if (absmax(r, m) < ftol) { status = 0; break; }
        jac_eval(c, z);
        njev++;
        double f = 0.5 * vdot(r, r, m);
        jt_mul(c, r, g);
        if (absmax(g, n) < gtol) { status = 2; break; }
        gn_step(c, r, g, p_gn);
        double gn_norm = vnorm(p_gn, n);
        if (gn_norm <= delta) {
            memcpy(p, p_gn, sizeof(double) * (size_t)n);
        } else {
            j_mul(c, g, tmp);
            double jg = vdot(tmp, tmp, m);
            double alpha = jg > 0.0 ? vdot(g, g, n) / jg : 0.0;
            for (int i = 0; i < n; i++) p_sd[i] = -alpha * g[i];
            double sd_norm = vnorm(p_sd, n);
            if (sd_norm >= delta) {
                double f2 = sd_norm > 0 ? delta / sd_norm : 0.0;
                for (int i = 0; i < n; i++) p[i] = p_sd[i] * f2;
            } else {
                double aa = 0.0, bb = 0.0;
                for (int i = 0; i < n; i++) { double d = p_gn[i] - p_sd[i]; aa += d * d; bb += 2.0 * p_sd[i] * d; }
                double cc = sd_norm * sd_norm - delta * delta;
                double disc = bb * bb - 4.0 * aa * cc;
                double tau = aa > 0.0 ? (-bb + sqrt(disc > 0 ? disc : 0.0)) / (2.0 * aa) : 0.0;
                for (int i = 0; i < n; i++) p[i] = p_sd[i] + tau * (p_gn[i] - p_sd[i]);
            }
        }
        double pnorm = vnorm(p, n);
        if (pnorm < xtol * (1.0 + vnorm(z, n))) { status = 1; break; }
        for (int i = 0; i < n; i++) z_new[i] = z[i] + p[i];
        res_eval(c, z_new, r_new);
        nfev++;
        double f_new = 0.5 * vdot(r_new, r_new, m);
        j_mul(c, p, tmp);
        double lin = 0.0;
        for (int i = 0; i < m; i++) { double v = r[i] + tmp[i]; lin += v * v; }
        double pred = f - 0.5 * lin;
        double rho = pred > 0.0 ? (f - f_new) / pred : (f_new < f ? 1.0 : -1.0);
        if (rho > 0.0) {
            memcpy(z, z_new, sizeof(double) * (size_t)n);
            memcpy(r, r_new, sizeof(double) * (size_t)m);
            if (rho > 0.75) { if (isfinite(delta)) delta = delta > 3 * pnorm ? delta : 3 * pnorm; }
            else if (rho < 0.25) delta = 0.5 * pnorm;
        } else {
            delta = 0.25 * pnorm;
        }
        if (delta < 1e-15 * (1.0 + vnorm(z, n))) { status = 3; it++; break; }
        if (nfev >= max_nfev) { status = 4; it++; break; }
    }
    info->status = status;
    info->nfev = nfev;
    info->njev = njev;
    info->iterations = it;
    info->rank = c->rank;
    free(g); free(p); free(p_gn); free(p_sd); free(z_new); free(r_new); free(tmp);
    return status;
}

/* -- Levenberg-Marquardt (Nielsen damping) ---------------------------------- */

static int lm_core(jac_ctx *c, double *z, double *r, double ftol, double xtol, double gtol, int max_iter, int max_nfev,
                   gcs_info *info)
{
    const double tau0 = 1e-8;
    int m = c->m, n = c->n;
    double *g = (double *)malloc(sizeof(double) * (size_t)n);
    double *p = (double *)malloc(sizeof(double) * (size_t)n);
    double *D = (double *)malloc(sizeof(double) * (size_t)n);
    double *damp = (double *)malloc(sizeof(double) * (size_t)n);
    double *z_new = (double *)malloc(sizeof(double) * (size_t)n);
    double *r_new = (double *)malloc(sizeof(double) * (size_t)(m ? m : 1));
    double *A = c->dense ? (double *)malloc(sizeof(double) * (size_t)n * n) : NULL;
    double *Ad = c->dense ? (double *)malloc(sizeof(double) * (size_t)n * n) : NULL;
    int nfev = 1, njev = 0, status = 4, it = 0;
    double lam = -1.0, nu = 2.0;
    for (it = 0; it < max_iter; it++) {
        if (absmax(r, m) < ftol) { status = 0; break; }
        jac_eval(c, z);
        njev++;
        jt_mul(c, r, g);
        if (absmax(g, n) < gtol) { status = 2; break; }
        if (c->dense) {
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++) {
                    double s = 0.0;
                    for (int k = 0; k < m; k++) s += c->J[(size_t)k * n + i] * c->J[(size_t)k * n + j];
                    A[(size_t)i * n + j] = s;
                }
            for (int i = 0; i < n; i++) D[i] = A[(size_t)i * n + i];
        } else {
            if (!c->s->ata) c->s->ata = gcs_ata_new(m, n, c->s->csr_indptr, c->s->csr_indices);
            gcs_ata_fill(c->s->ata, c->data);
            gcs_ata_diag(c->s->ata, D);
        }
        double dmax = 0.0;
        for (int i = 0; i < n; i++) if (D[i] > dmax) dmax = D[i];
        double floor_ = dmax > 0 ? 1e-8 * dmax : 1e-8;
        for (int i = 0; i < n; i++) if (D[i] < floor_) D[i] = floor_;
        if (lam < 0.0) lam = tau0 * (dmax > 0 ? dmax : 1.0);
        double f = 0.5 * vdot(r, r, m);
        for (;;) {
            for (int i = 0; i < n; i++) damp[i] = lam * D[i];
            int bad = 0;
            if (c->dense) {
                memcpy(Ad, A, sizeof(double) * (size_t)n * n);
                for (int i = 0; i < n; i++) Ad[(size_t)i * n + i] += damp[i];
                for (int i = 0; i < n; i++) p[i] = -g[i];
                bad = gcs_lu_solve(n, Ad, p) != 0;
            } else {
                for (int i = 0; i < n; i++) p[i] = -g[i];
                bad = gcs_ata_solve(c->s->ata, damp, p) != 0;
            }
            if (bad) { lam *= nu; nu *= 2; if (lam > 1e32) { status = 3; goto done; } continue; }
            double pnorm = vnorm(p, n);
            if (pnorm < xtol * (1.0 + vnorm(z, n))) { status = 1; goto done; }
            for (int i = 0; i < n; i++) z_new[i] = z[i] + p[i];
            res_eval(c, z_new, r_new);
            nfev++;
            double f_new = 0.5 * vdot(r_new, r_new, m);
            double pred = 0.0;
            for (int i = 0; i < n; i++) pred += p[i] * (damp[i] * p[i] - g[i]);
            pred *= 0.5;
            double rho = pred > 0.0 ? (f - f_new) / pred : -1.0;
            if (rho > 0.0) {
                memcpy(z, z_new, sizeof(double) * (size_t)n);
                memcpy(r, r_new, sizeof(double) * (size_t)m);
                double t = 1.0 - pow(2.0 * rho - 1.0, 3.0);
                lam *= (t > 1.0 / 3.0 ? t : 1.0 / 3.0);
                nu = 2.0;
                break;
            }
            lam *= nu;
            nu *= 2;
            if (nfev >= max_nfev || lam > 1e32) { status = nfev >= max_nfev ? 4 : 3; it++; goto done; }
        }
    }
done:
    info->status = status;
    info->nfev = nfev;
    info->njev = njev;
    info->iterations = it;
    info->rank = -1;
    free(g); free(p); free(D); free(damp); free(z_new); free(r_new); free(A); free(Ad);
    return status;
}

/* -- entry points ------------------------------------------------------------ */

int gcs_system_solve(gcs_system *s, int method, double ftol, double xtol, double gtol,
                     int max_iter, int max_nfev, int dense, double *z, gcs_info *info)
{
    info->status = 0; info->nfev = 0; info->njev = 0; info->iterations = 0; info->rank = -1;
    if (s->n_free == 0 || s->n_res == 0) { info->nfev = 1; info->status = 0; return 0; }
    if (dense < 0) dense = s->n_free <= 120;
    if (max_nfev <= 0) max_nfev = 4 * max_iter;
    jac_ctx c;
    memset(&c, 0, sizeof(c));
    c.s = s; c.dense = dense; c.m = s->n_res; c.n = s->n_free;
    c.J = dense ? (double *)malloc(sizeof(double) * (size_t)s->n_res * s->n_free) : NULL;
    c.data = s->csr_data;
    c.work = dense ? NULL : (double *)malloc(sizeof(double) * (size_t)s->n_free);
    double *r = (double *)malloc(sizeof(double) * (size_t)s->n_res);
    gcs_system_residuals(s, z, r);
    int st = (method == GCS_LM)
        ? lm_core(&c, z, r, ftol, xtol, gtol, max_iter, max_nfev, info)
        : dogleg_core(&c, z, r, ftol, xtol, gtol, max_iter, max_nfev, info);
    gcs_system_apply_z(s, z);
    free(r); free(c.work); free(c.J);
    return st;
}
